### PR TITLE
[Feature] Replace feature change notifications

### PR DESCRIPTION
**Pull request by KaterinaWire**
## What's new in this PR?

1. Add observer for feature change.
2. Refactor the feature change notifications.

### CODE REVIEW

File: Tests/Source/Model/Observer/ObjectObserverToken/FeatureObserverTests.swift [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        XCTAssertEqual(observer.notifications.count, changeCount, "Should not have changed further once")
+
+        guard let changes = observer.notifications.first else { return }
+        changes.checkForExpectedChangeFields(userInfoKeys: userInfoKeys,
+                                             expectedChangedFields: expectedChangedFields,
+                                             file: file,
+                                             line: line)
+    }
+
+    func testThatItNotifiesTheObserverOfChangedStatus() {
```
Comment (johnxnguyen) [inline Tests/Source/Model/Observer/ObjectObserverToken/FeatureObserverTests.swift:94]: Will we also test the config changed?

**Review by typfel — APPROVED** [review]

File: Tests/Source/Model/Observer/ObjectObserverToken/FeatureObserverTests.swift (hunk 1 above)
Comment (KaterinaWire) [reply Tests/Source/Model/Observer/ObjectObserverToken/FeatureObserverTests.swift:94]: Yes, added.

**Review by johnxnguyen — APPROVED** [review]

File: Source/Model/FeatureConfig/FeatureService.swift [hunk 2]
```diff
@@ -143,7 +140,7 @@ public class FeatureService {
         return result
     }
 
-    // Maybe call this acknowledgeChanges
+    // Call this to acknowledge changes
```
[the remark this answers is not in the document]
Comment (KaterinaWire) [reply Source/Model/FeatureConfig/FeatureService.swift:143]: We need this method to acknowledge feature changes from the UI

**Comment by KaterinaWire** [comment]
> So previously there was `addObserver()` on the Feature entity and that has now been replaced by a `Notification`? Can I get some context on what problem this is solving?

@typfel The original idea was to use CoreData notification to notify users about all feature changes. But with @johnxnguyen, we've come to the conclusion that it might be better at this point to implement feature change notification using Notification. It's because the first time the app is launched, all observable keys are set to changed and the user is notified of notifications which were not actually made. This case is not an issue when we update the UI, but it's an issue when we need to display an alert.

File: Source/Model/FeatureConfig/FeatureService.swift [hunk 3]
```diff
@@ -34,14 +34,11 @@ public class FeatureService {
...
 
     private let context: NSManagedObjectContext
-    private var observerToken: Any?
-    public weak var delegate: FeatureServiceDelegate?
 
     // MARK: - Life cycle
 
     public init(context: NSManagedObjectContext) {
         self.context = context
-        self.observerToken = Feature.addObserver(self, in: context)
```
Comment (KaterinaWire) [inline Source/Model/FeatureConfig/FeatureService.swift:44]: @johnxnguyen Here's the place where we used the observer.

**Comment by KaterinaWire** [comment]
> I remember there is code to add the observer for `FeatureChangeInfo` or something like that. That would need to be deleted too.

@johnxnguyen Yes, I pointed to this code above

File: Source/Model/FeatureConfig/FeatureService.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (johnxnguyen) [reply Source/Model/FeatureConfig/FeatureService.swift:143]: I'm referring to the comment (it was a note to myself to possibly rename the method).

[the remark this answers is not in the document]
Comment (KaterinaWire) [reply Source/Model/FeatureConfig/FeatureService.swift]: I removed the comment. I think we can keep this name (`setNeedsToNotifyUser(_ notifyUser: Bool, for featureName: Feature.Name)`)

**Review by typfel — APPROVED** [review]

**Pull request merged**